shiny/ui/_theme.py

```diff
@@ -0,0 +1,534 @@
+        self._reset_css()
+        return self
+
+    def add_rules(
+        self: T,
+        *args: str,
+        **kwargs: str | float | int | bool | None,
+    ) -> T:
+        """
+        Add custom Sass rules to the theme.
+
+        Sass code added via this method will be placed **after** the rule declarations
+        from the theme preset, allowing you to override or extend the default rules.
+
+        Parameters
+        ----------
+        *args
+            Sass code, as a single or multiple strings, containing rules to add.
+        **kwargs
+            Keyword arguments containing Sass value declarations to add. The keys
+            should be Sass variable names using underscore casing that will be
+            transformed automatically to kebab-case. For example,
+            `.add_rules(primary_color="#ff0000")` is equivalent to
+            `.add_rules("$primary-color: #ff0000;")`.
+        """
+        rules = self._combine_args_kwargs(*args, kwargs=kwargs)
+        self._rules.extend(rules)
+        self._reset_css()
+        return self
+
+    def to_sass(self) -> str:
+        """
+        Returns the custom theme as a single Sass string.
+
+        Returns
+        -------
+        :
+            The custom theme as a single Sass string.
+        """
+        path_functions = path_pkg_preset(self._preset, "_01_functions.scss")
+        path_defaults = path_pkg_preset(self._preset, "_02_defaults.scss")
+        path_mixins = path_pkg_preset(self._preset, "_03_mixins.scss")
+        path_rules = path_pkg_preset(self._preset, "_04_rules.scss")
+
+        sass_lines = [
+            f'@import "{path_functions}";',
+            *self._functions,
+            *self._defaults,
+            f'@import "{path_defaults}";',
+            f'@import "{path_mixins}";',
+            *self._mixins,
+            f'@import "{path_rules}";',
+            *self._rules,
+        ]
+
+        return "\n".join(sass_lines)
+
+    def to_css(
+        self,
+        compile_args: Optional[SassCompileArgs] = None,
+    ) -> str:
+        """
+        Compile the theme to CSS and return the result as a string.
+
+        Parameters
+        ----------
+        compile_args
+            A dictionary of keyword arguments to pass to
+            [`sass.compile()`](https://sass.github.io/libsass-python/sass.html#sass.compile).
+
+        Returns
+        -------
+        :
+            The compiled CSS for the theme. The value is cached such that previously
+            compiled themes are returned immediately. Adding additional custom Sass code
+            or changing the preset will invalidate the cache.
+        """
+        if self._css:
+            return self._css
+
+        if self._can_use_precompiled():
+            self._css = self._read_precompiled_css()
+            return self._css
+
+        check_libsass_installed()
+        import sass
+
+        if compile_args is None:
+            compile_args = {"output_style": "compressed"}
```

Comment:
This way `output_style`'s default won't change if you specify some other option 

```suggestion
        if compile_args is None:
            compile_args = {}
        compile_args = {"output_style": "compressed", **compile_args}
```

Reply:
Please ignore this suggestion and see this comment (https://github.com/posit-dev/py-shiny/pull/1358#issuecomment-2192454906) instead

Reply:
Great catch and thanks for the changes in the patch!